File: src/main/java/com/oycm/algorithm/ba/opposite/Solution_14.scala
package com.oycm.algorithm.ba.opposite

object Solution_14 {

  /**
   * LCP 28. 采购方案
   * https://leetcode.cn/problems/4xy4Wx/description/
   *
   * 小力仅购买两个零件，要求购买零件的花费不超过预算，请问他有多少种采购方案。
   *
   * @param nums   n 个零件的报价, [1, 10 pow 5]
   * @param target 小力预算
   * @return
   */
  def purchasePlans(nums: Array[Int], target: Int): Int = {
    /*
    对数组进行升序排序，在数组中找 nums(i) + nums(j) <= target 的组合数
    i < j, 如果 nums(i) + nums(j) <= target, j - i 中的组合都符合答案
    l = 0, r = n - 1; sum > target, 则 r--, 当 sum <= target 时, 记录答案 r-l, l++
    l 和 r 不断靠近
    时间复杂度在排序 n log n
     */
    val sort = nums.sorted
    var ans = 0L
    var l = 0
    var r = nums.length - 1
    while (l < r && sort(l) < target) {
      while (l < r && sort(l) + sort(r) > target) {
        r -= 1
      }
      ans += r - l

      l += 1
    }

    (ans % 1000000007).toInt
  }

  def main(args: Array[String]): Unit = {
    println(purchasePlans(Array(2, 5, 3, 5), 6))
    println(purchasePlans(Array(2, 2, 1, 9), 3))
  }

}

File: src/main/java/com/oycm/algorithm/h/basic/Solution_14.java
package com.oycm.algorithm.h.basic;

public class Solution_14 {

    /**
     * 231. <a href="https://leetcode.cn/problems/power-of-two/description/">2 的幂</a>
     *
     * @param n 有正整数和负整数
     * @return
     */
    public boolean isPowerOfTwo(int n) {
        /*
        n < 0 不是 2 的幂次
        1 的 个数
        或
        n & (n-1) == 0
         */
        return n > 0 && Integer.bitCount(n) == 1;
    }
}
